doc_indexer/src/indexing/indexer.py

```diff
@@ -83,47 +88,180 @@ def _load_documents(self) -> list[Document]:
+        self._handle_database_operation(
+            operation, only_warn_if_table_inexistent=only_warn_if_table_inexistent
+        )
+
+    def _index_chunks_in_batches(self, chunks: list[Document]) -> None:
+        """
+        Indexes and stores document chunks in batches to the temporary HanaDB table.
+
+        Args:
+            chunks (list[Document]): List of document chunks to index and store.
+        """
         logger.info("Indexing and storing indexes to HanaDB...")
-        for i in range(0, len(all_chunks), CHUNKS_BATCH_SIZE):
-            batch = all_chunks[i : i + CHUNKS_BATCH_SIZE]
+        for i in range(0, len(chunks), CHUNKS_BATCH_SIZE):
+            batch = chunks[i : i + CHUNKS_BATCH_SIZE]
             try:
-                # Add current batch of documents
                 self.db.add_documents(batch)
                 logger.info(
-                    f"Indexed batch {i//CHUNKS_BATCH_SIZE + 1} of {len(all_chunks)//CHUNKS_BATCH_SIZE + 1}"
+                    f"Indexed batch {i // CHUNKS_BATCH_SIZE + 1} of {len(chunks) // CHUNKS_BATCH_SIZE + 1}"
                 )
-
-                # Wait 3 seconds before processing next batch
-                if i + CHUNKS_BATCH_SIZE < len(all_chunks):
+                if i + CHUNKS_BATCH_SIZE < len(chunks):
                     time.sleep(3)
-
-            except Exception:
-                logger.exception(
-                    f"Error while storing documents batch {i//CHUNKS_BATCH_SIZE + 1} in HanaDB"
-                )
+            except Exception as e:
+                logger.error(f"Batch {i // CHUNKS_BATCH_SIZE + 1} failed: {e}")
                 raise
+        logger.info(
+            f"Successfully indexed {len(chunks)} markdown files chunks to {self.db.table_name} table."
+        )
+
+    def index(self) -> None:
+        """
+        Indexes markdown files in the specified directory by chunking them and storing the chunks in the database.
+
+        The process involves:
+            1. Loading and chunking markdown documents by specified headers.
+            2. Backing up the current main table (if it exists).
+            3. Deleting all data from the main table.
+            4. Storing new document chunks in the main table.
+            5. On error, restoring the main table from backup.
+
+        Raises:
+            Exception: If indexing fails and restoration from backup is attempted.
+        """
+
+        docs = self._load_documents()
+
+        # Chunk the documents by the headers.
+        chunks = create_chunks(docs, self.headers_to_split_on)
+
+        # Create a backup of the current main table (if exists).
+        logger.info(
+            f"Backing up current table {self.db.table_name} to {self.backup_table_name} if exists..."
+        )
+        self._copy_table(
+            self.db.table_name,
+            self.backup_table_name,
+            only_warn_if_table_inexistent=True,
+        )
```

Comment:
what if copying the table fails? Does it stop proceeding here?

Reply:
It depends: 
If the table we are trying to copy does not exist, we only will log a warning (`only_warn_if_table_inexistent=True`).
If anything else happens, the process stops here.